File: src/spark-examples/sparknormal-examples/SparkStreamingJavaExample/src/com/huawei/bigdata/spark/examples/FemaleInfoCollectionPrint.java
package com.huawei.bigdata.spark.examples;

import kafka.serializer.StringDecoder;
import scala.Tuple2;
import scala.Tuple3;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.streaming.Duration;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.streaming.kafka.KafkaUtils;

import java.util.*;

/**
 * Consumes messages from one or more topics in Kafka.
 * <batchTime> is the Spark Streaming batch duration in seconds.
 * <topics> is a list of one or more kafka topics to consume from
 * <brokers> is for bootstrapping and the producer will only use it for getting metadata
 */
public class FemaleInfoCollectionPrint {
    public static void main(String[] args) throws Exception {
        String checkPointDir = args[0];
        String batchTime = args[1];
        String topics = args[2];
        String brokers = args[3];

        Duration batchDuration = Durations.seconds(Integer.parseInt(batchTime));

        // Create a Streaming startup environment.
        SparkConf conf = new SparkConf().setAppName("DataSightStreamingExample");
        JavaStreamingContext jssc = new JavaStreamingContext(conf, batchDuration);

        // Configure the CheckPoint directory for the Streaming.
        jssc.checkpoint(checkPointDir);

        // Get the list of topic used by kafka
        HashSet<String> topicsSet = new HashSet<String>(Arrays.asList(topics.split(",")));
        HashMap<String, String> kafkaParams = new HashMap<String, String>();
        kafkaParams.put("metadata.broker.list", brokers);

        // Create direct kafka stream with brokers and topics
        // Receive data from the Kafka and generate the corresponding DStream
        JavaDStream<String> lines =
                KafkaUtils.createDirectStream(
                                jssc,
                                String.class,
                                String.class,
                                StringDecoder.class,
                                StringDecoder.class,
                                kafkaParams,
                                topicsSet)
                        .map(
                                new Function<Tuple2<String, String>, String>() {
                                    public String call(Tuple2<String, String> tuple2) {
                                        return tuple2._2();
                                    }
                                });

        // Obtain field properties in each row.
        JavaDStream<Tuple3<String, String, Integer>> records =
                lines.map(
                        new Function<String, Tuple3<String, String, Integer>>() {
                            public Tuple3<String, String, Integer> call(String line) throws Exception {
                                String[] elems = line.split(",");
                                return new Tuple3<String, String, Integer>(
                                        elems[0], elems[1], Integer.parseInt(elems[2]));
                            }
                        });

        // Filter data about the time that female netizens spend online
        JavaDStream<Tuple2<String, Integer>> femaleRecords =
                records.filter(
                                new Function<Tuple3<String, String, Integer>, Boolean>() {
                                    public Boolean call(Tuple3<String, String, Integer> line) throws Exception {
                                        if (line._2().equals("female")) {
                                            return true;
                                        } else {
                                            return false;
                                        }
                                    }
                                })
                        .map(
                                new Function<Tuple3<String, String, Integer>, Tuple2<String, Integer>>() {
                                    public Tuple2<String, Integer> call(
                                            Tuple3<String, String, Integer> stringStringIntegerTuple3)
                                            throws Exception {
                                        return new Tuple2<String, Integer>(
                                                stringStringIntegerTuple3._1(), stringStringIntegerTuple3._3());
                                    }
                                });

        // Filter data about users whose consecutive online duration exceeds the threshold.
        JavaDStream<Tuple2<String, Integer>> upTimeUser =
                femaleRecords.filter(
                        new Function<Tuple2<String, Integer>, Boolean>() {
                            public Boolean call(Tuple2<String, Integer> stringIntegerTuple2) throws Exception {
                                if (stringIntegerTuple2._2() > 30) {
                                    return true;
                                } else {
                                    return false;
                                }
                            }
                        });

        // print the results
        upTimeUser.print();

        // The Streaming system starts.
        jssc.start();
        jssc.awaitTermination();
    }
}

File: src/spark-examples/sparksecurity-examples/SparkStreamingPythonExample/FemaleInfoCollectionPrint.py
# -*- coding:utf-8 -*-
# kafka 0.8
import sys
from pyspark.streaming import StreamingContext, kafka
from pyspark import SparkConf, SparkContext


if __name__ == "__main__":
    if len(sys.argv) < 5:
        print("Usage: FemaleInfoCollectionPrint.py <checkPointDir> <batchTime> <topics> <brokers>")
        exit(-1)

    checkPointDir = sys.argv[1]
    batchTime = sys.argv[2]
    topics = sys.argv[3]
    brokers = sys.argv[4]

    # 初始化SparkContext
    conf = SparkConf().setAppName("DataSightStreamingExample")
    sc = SparkContext(conf = conf)
    ssc = StreamingContext(sc, int(batchTime))

    # 设置Streaming的CheckPoint目录
    ssc.checkpoint(checkPointDir)

    # 获得Kafka的主题和brokers列表
    topics = list(set(topics.split(",")))
    kafkaParams = {"metadata.broker.list": brokers}

    # 通过brokers和topics直接创建kafka stream
    # 1.接收Kafka中数据，生成相应DStream
    lines = kafka.KafkaUtils.createDirectStream(ssc, topics, kafkaParams).map(lambda r:r[1])
	
    # 2.获取每一个行的字段属性
    records = lines.map(lambda line: line.split(",")).map(lambda elems: (elems[0], elems[1], int(elems[2])))

    # 3.筛选女性网民上网时间数据信息
    femaleRecords = records.filter(lambda record1: record1[1] == "female").map(lambda record2: (record2[0], record2[2]))
    upTimeUser = femaleRecords.filter(lambda User: User[1] > 30)

    # 4.筛选连续上网时间超过阈值的用户，并获取结果
    # 默认打印DStream中前10个元素
    upTimeUser.pprint()

    # 5.Streaming系统启动
    ssc.start()
    ssc.awaitTermination()
